reimplement debugInfos workflow

PPpro opened the pull request:
Re: #2384 

Changelog:
 * 

<!-- Note: Makes sure these boxes are checked before submitting your PR - thank you!

- [ ] If your pull request has gone "stale", you should **rebase** your work on top of the latest version of the upstream branch.
- [ ] If your commit history is full of small, unimportant commits (such as "fix pep8" or "update tests"), **squash** your commits down to a few, or one, discreet changesets before submitting a pull request.

- To official teams:
  - [ ] Check that your javascript is following our [style guide](https://github.com/cocos-creator/fireball/blob/dev/.github/CONTRIBUTING.md) and end files with a newline
  - [ ] Document new code with comments in source code based on [API Docs](https://github.com/cocos-creator/fireball#api-docs)
  - [ ] Make sure any runtime log information in `cc.log` , `cc.error` or `new Error('')` has been moved into `DebugInfos.js` with an ID, and use `cc.logID(id)` or `new Error(cc._getErrorID(id))` instead.

-->

jareguo commented on EngineErrorMap.md:
````diff
@@ -1096,6 +1231,58 @@ Unknown 'type' parameter of %s.%s：%s
+@property(...)
+value = 0
+```
+
````
3655 漏了...

jareguo commented on gulp/tasks/engine.js:
```diff
@@ -315,4 +321,30 @@ exports.buildJsbMin = function (sourceFile, outputFile, excludes, opt_macroFlags
+
+    let writeContent = `${JSON.stringify(debugInfos, null, 4)}`;
+    Fs.writeFileSync('DebugInfos.json', writeContent);
+};
```
这个 task 单独放在一个 js 中吧

jareguo commented on gulp/tasks/engine.js:
```diff
@@ -315,4 +321,30 @@ exports.buildJsbMin = function (sourceFile, outputFile, excludes, opt_macroFlags
+        result1 = result2;
+    }
+
+    let writeContent = `${JSON.stringify(debugInfos, null, 4)}`;
```
这个模板字符串用的好像没意义啊

jareguo commented on DebugInfos.json:
```diff
@@ -0,0 +1,406 @@
+{
```
这个文件不需要提交到 repo，加到 .gitignore 吧

jareguo commented on gulpfile.js:
```diff
@@ -49,6 +49,7 @@ gulp.task('build-html5-min', function (done) {
 
 gulp.task('build-html5-preview',  function (done) {
     Engine.buildPreview('./index.js', './bin/cocos2d-js-for-preview.js', done);
+    Engine.buildDebugInfos();
```
应该是改成 gulp.task('build-html5-preview',  'build-debug-infos', function...
还有其它 build-xx 也需要依赖 build-debug-infos

jareguo commented on README.md:
````diff
@@ -80,6 +80,57 @@ npm test
+          
+    ```
+
+2. Define deprecated log
````
讲一下什么时候需要标记为 deprecated

jareguo commented on README.md:
```diff
@@ -80,6 +80,57 @@ npm test
+
+For details below:
+
+1. Define log  
```
Define log in EngineErrorMap.md

jareguo commented on README.md:
````diff
@@ -80,6 +80,57 @@ npm test
+    ```
+3. Attentions
+    
+    You cannot leave blank line at the end-of-file
````
这是为什么呀？

PPpro replied:
因为解析的时候， 文件尾那一句 我是截取到文件尾的
如果有空行，可能空行也会被解析进去哦，比如：
"8400": "Wrong type arguments, 'filePath' must be a String.\n\n",

这里好像也没什么好的办法判断最后一句的结尾位置在哪，所以我就截取到文件尾了
或者要再md 的文件尾 做个标记吗
<!-- END-OF-FILE -->

PPpro replied:
<!- - END-OF-FILE -- >

jareguo replied:
trimRight

pandamicro commented on README.md:
```diff
@@ -80,6 +80,57 @@ npm test
 
+### DebugInfos  
+
+All the debug infos are defined in file EngineErrorMap.md.  
```
给 EngineErrorMap.md 加个链接

pandamicro commented on EngineErrorMap.md:
```diff
@@ -1816,16 +2264,83 @@ cc.TMXMapInfo.parseXMLFile(): unsupported compression method
+
 ### 7300
 
+<!-- DEPRECATED -->
```
这些 DEPRECATED 是怎么出来的？人工判断？

PPpro replied:
对的，纯人工复查

jareguo replied:
就是存档用的，给旧版本用户查文档的时候看的。新版本用不到这些 log 了

pandamicro commented on CCDebugger.js:
```diff
@@ -310,7 +310,7 @@ cc._throw = CC_EDITOR ? Editor.error : function (error) {
 // define log methods to lookup message ID
 
-const debugInfos = CC_DEBUG && require('./DebugInfos');
+const debugInfos = require('./DebugInfos.json') || {};
```
可以直接 require json ？？？

jareguo replied:
可以

2youyou2 replied:
指定 .json 2.0会有问题的额，2.0 会把 .json 转成 .js

jareguo replied:
json 的支持好加吗？

2youyou2 replied:
不用加 .json 后缀，`require('./DebugInfos')` 就好

jareguo replied:
转成 js 后，2.0 会自动加上 module.exports = JSON 吗？

2youyou2 replied:
对的

PPpro replied:
已经提交修改

PPpro committed:
reimplement debugInfos workflow


debugInfos


debuginfos


README updated


debugInfos workflow


reimplement debugInfos workflow

PPpro committed:
add error id

PPpro committed:
Merge remote-tracking branch 'origin/master'

PPpro committed:
merge gulpfile
